src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs

```diff
@@ -1701,119 +1672,355 @@ bool isRefEscape
+        {
+            if (receiver is { })
             {
-                escapeTo = GetValEscape(receiverOpt, scopeOfTheContainingExpression);
+                argsAndParams.Add(getReceiver(symbol, receiver));
             }
 
             if (!argsOpt.IsDefault)
             {
-                BoundArgListOperator? argList = null;
-                for (var argIndex = 0; argIndex < argsOpt.Length; argIndex++)
+                for (int argIndex = 0; argIndex < argsOpt.Length; argIndex++)
                 {
                     var argument = argsOpt[argIndex];
                     if (argument.Kind == BoundKind.ArgListOperator)
                     {
-                        Debug.Assert(argIndex == argsOpt.Length - 1, "vararg must be the last");
-                        argList = (BoundArgListOperator)argument;
+                        Debug.Assert(argIndex == argsOpt.Length - 1);
+                        // unwrap varargs and process as more arguments
+                        var argList = (BoundArgListOperator)argument;
+                        getArgList(
+                            argList.Arguments,
+                            ignoreArglistRefKinds ? default : argList.ArgumentRefKindsOpt,
+                            argsAndParams);
                         break;
                     }
 
-                    var parameter = getParameter(parameters, argsToParamsOpt, argIndex);
-                    if (!considerParameter(parameter))
+                    var parameter = argIndex < parameters.Length ?
+                        parameters[argsToParamsOpt.IsDefault ? argIndex : argsToParamsOpt[argIndex]] :
+                        null;
+
+                    var refKind = parameter?.RefKind ?? RefKind.None;
+                    if (!argRefKindsOpt.IsDefault)
                     {
-                        continue;
+                        refKind = argRefKindsOpt[argIndex];
+                        if (refKind == RefKind.None &&
+                            parameter?.RefKind == RefKind.In)
+                        {
+                            refKind = RefKind.In;
+                        }
                     }
 
-                    updateEscapeTo(argument, parameter.RefKind, scopeOfTheContainingExpression, ref escapeTo);
+                    argsAndParams.Add((parameter, argument, refKind));
                 }
+            }
 
-                if (argList != null)
+            static (ParameterSymbol? Parameter, BoundExpression Argument, RefKind RefKind) getReceiver(Symbol symbol, BoundExpression receiver)
+            {
+                if (symbol is FunctionPointerMethodSymbol)
                 {
-                    var argListArgs = argList.Arguments;
-                    var argListRefKindsOpt = argList.ArgumentRefKindsOpt;
+                    return (null, receiver, RefKind.None);
+                }
+                var containingType = symbol.ContainingType;
+                var refKind = containingType.IsStructType() ?
```

Comment:
I think I agree we should use `method?.ThisParameter.RefKind ?? RefKind.None`. For a constructor, `ThisParameter.RefKind == RefKind.Out` though, rather than `RefKind.None` which is used here, so we'll need to investigate the impact.